keps/sig-network/20200103-mixed-protocol-lb.md

```diff
@@ -0,0 +1,392 @@
+
+Our feature does not introduce new values or new fields. It enables the usage of an existing value in existing fields, but with a different logic. I.e. if someone creates a Service with mixed protocol setup and then rollbacks the API server to a version that does not implement this feature the clients will still get the Service with mixed protocols when they read that via the rollback'd API. If the client (CPI implementation) has been rollbacked, too, then the client may receive such a Service setup that it does not support.
+
+- Alibaba: no risk. The current CPI and LB already supports the mixed protocols in the same Service definition. If this feature is enabled in an API server and then the API server rollback is executed the CPI can still handle the Services with mixed protocol sets.
+- AWS: no risk. The current CPI accepts Services with TCP protocol only, i.e. after a K8s upgrade a user still cannot use this feature. Consequently, a rollback in the K8s version does not introduce any issues.
+- Azure: no risk. The current CPI and LB already supports the mixed protocols in the same Service definition. The situation is the same as with the Alibaba CPI.
+- GCE: currently the GCE CPI assumes that a Service definition contains a single protocol value, as it assumes that the Service Controller already rejected Services with mixed protocols. While the Service Controller really did so a while ago, it does not do this anymore. It means a risk.
+- IBM Cloud VPC: no risk. The same situation like in the case of AWS.
+- IBM Cloud Classic: no risk. The CPI and NLB already supports TCP and UDP in the same Service definition. The same situation like in the case of Alibaba.
+- OpenStack: no risk. The CPI and NLB already supports TCP and UDP in the same Service definition. The same situation like in the case of Alibaba.
+- Oracle: no risk. The CPI and LB already supports mixed protocols. The same situation like in the case of Alibaba.
+- Tencent Cloud: no risk. The CPI and LB already supports mixed protocols. The same situation like in the case of Alibaba.
+
+
+The other risk is in the introduction of this feature without an option control mechanism, i.e. as a general change in Service handling. In that case there is the question whether this feature should be a part of the conformance test set, because it can affect the conformance of cloud providers.
+
+A possible mitigation is to put the feature behind option control. 
+
+## Design Details
+
+The implementation of the basic logic is ready in this PR:
+https://github.com/kubernetes/kubernetes/pull/75831
+
+Currently a feature gate is used to control its activation status. Though if we want to keep this feature behind option control even after it reaches its GA state we should come up with a different solution, as feature gates are used to control the status of a feature as that graduates from alpha to GA, and they are not meant for option control for features with GA status.
```

Comment:
I'm not sure an alpha is warranted here, but maybe.  It's not a huge change, except in the cloud providers it could be.  Thoughts?

Reply:
Yeah I think we skip alpha as long as rollbacks don't break LBs as you mentioned here https://github.com/kubernetes/enhancements/pull/1438#discussion_r367675751

Reply:
Do I understand right then that the agreed change is the following:
- in 1.18 creating a mixed Service is allowed if the feature flag is ON. We add notification to the documentation that in this case rollback is not possible, or at least can break things depending on the cloud provider implementation
- in 1.18 we allow update of such Services even if the feature flag in OFF
- in 1.19 we change the create path so, that even if the feature flag is OFF mixed Services can be created
?
Thank you!

Reply:
> in 1.18 creating a mixed Service is allowed if the feature flag is ON. We add notification to the documentation that in this case rollback is not possible, or at least can break things depending on the cloud provider implementation

That's correct, and that means the feature gate should be alpha and disabled by default (we should not enable something by default that breaks skew/rollback).

> in 1.18 we allow update of such Services even if the feature flag in OFF

Correct

> in 1.19 we change the create path so, that even if the feature flag is OFF mixed Services can be created
?

Typically, we would leave the create/update logic the same as in 1.18, and simply promote the feature flag to beta and enable it by default in 1.19. That gives a release to get broader feedback and bug reports, but allows someone to turn off the feature and stop allowing new uses if problems are found, without cutting a new Kubernetes 1.19.x patch release.

Once the feature graduates to GA and is locked on (e.g. in 1.20), *then* all feature-gate-conditional logic can be dropped.

Reply:
@liggitt @thockin @andrewsykim I think I updated the document according to this logic. Please check and suggest how to continue.